Add "Shared with you" badge to Extensions > My Subscriptions list as applicable

Dan-Q opened the pull request:
### Changes proposed in this Pull Request:

Fixes [WCCOM#19063](https://github.com/Automattic/woocommerce.com/issues/19063).

To make it easier to see when an extension is available because of a shared subscription, this PR amends the Extensions > My Subscriptions page for connected stores to:

- Add a "Shared with you" badge to products available as a result of a shared subscription
    - In a popup, shows the email address of the person who shared the subscription with you and a link to "learn more"
- Removes the "Manage on Woo.com" link from the actions menu for subscriptions shared with you
- Reduces the font size on such badges to 12px, in accordance with the latest designs
- Improves the spacing of these badges, where multiple badges appear on multiple lines, so they don't collide

![Screenshot highlighting the new badge and edited action menu.](https://github.com/woocommerce/woocommerce/assets/53293/24856ee4-afa6-477e-9d2b-9c77792aa295)

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. Connect a store using a Woo.com account that has at least one subscription shared with it.
2. In WooCommerce > Extensions > My Subscriptions, observe the new "Shared with you" badge.
3. Click on the badge to see the email address of the person sharing it and a link to the sharing documentation.
4. Note that the font size of the badge is now 12px.
5. Click the kebab menu menu and observe that the "Manage on Woo.com" link does not appear for shared subscriptions.
6. (Optional) Force a second badge to appear, e.g. by manually editing `plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/table/rows/functions.tsx` and adding another `<StatusPopover>`; observe that the badges have both horizontal _and now vertical_ margins to prevent them from colliding with one another.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

For shared subscriptions, My Subscriptions now shows "Shared with you" and the email address of the person who shared it with you.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Dan-Q committed:
Add 'shared with you' badge

Dan-Q committed:
Add 'shared with you' badge

Dan-Q committed:
Improve spacing where multiple badges span multiple lines

Dan-Q committed:
Improve spacing where multiple badges span multiple lines

Dan-Q committed:
Decrease font size of badges per new designs

Dan-Q committed:
Decrease font size of badges per new designs

Dan-Q committed:
Include 'Manage on Woo.com' only for non-shared subscriptions

Dan-Q committed:
Include 'Manage on Woo.com' only for non-shared subscriptions

Dan-Q committed:
Made TypeScript happier

Dan-Q committed:
Made TypeScript happier

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

andfinally committed:
Addressing linter errors.

andfinally reviewed (commented):
This is a brilliant improvement, thanks! 👌 

I wonder if there's anything we can do about the tooltip width? With my (admittedly unnaturally long) test email address, I get an ugly scrollbar.

Also, I just noticed the behaviour of this and the tooltip on the Expired badge is a bit annoying. It won't go away till you click the badge again. This is not very intuitive, IMHO. I notice we use the same component for the kebab menu, but I think we add an outside click handler when we show it, so you can close it by clicking anywhere outside it.

Dan-Q committed:
Merge branch 'trunk' into add/wccom19063-in-app-shared-by-badge

Dan-Q committed:
Merge branch 'add/wccom19063-in-app-shared-by-badge' of https://github.com/woocommerce/woocommerce into add/wccom19063-in-app-shared-by-badge

Dan-Q commented:
> I wonder if there's anything we can do about the tooltip width? With my (admittedly unnaturally long) test email address, I get an ugly scrollbar.

We can't change the width of the tooltip to anything sensible without breaking all kinds of other bits of design. How about we just make it so that exceptionally-long email addresses are allowed to wrap. It's the least-bad option, and only requires a single `word-wrap: anywhere;` CSS directive:

https://github.com/woocommerce/woocommerce/assets/53293/2870d79c-8905-4fad-9865-c0aa45311610

Dan-Q commented:
> Also, I just noticed the behaviour of this and the tooltip on the Expired badge is a bit annoying. It won't go away till you click the badge again. This is not very intuitive, IMHO. I notice we use the same component for the kebab menu, but I think we add an outside click handler when we show it, so you can close it by clicking anywhere outside it.

Suggest this is treated as a separate issue, as (a) it applies to all the `<StatusPopover>`s everwhere, and (b) should probably be fixed centrally for them all, which is kinda out-of-scope for this issue.

Dan-Q committed:
Allow unusually-long email addresses of sharers to wrap

Dan-Q commented:
> > I wonder if there's anything we can do about the tooltip width? With my (admittedly unnaturally long) test email address, I get an ugly scrollbar.
> 
> We can't change the width of the tooltip to anything sensible without breaking all kinds of other bits of design. How about we just make it so that exceptionally-long email addresses are allowed to wrap. It's the least-bad option, and only requires a single `word-wrap: anywhere;` CSS directive:

Done in e636709c204dbf8a8ed2ca273e37371f41f17271.

andfinally reviewed (commented):
Thanks, looks good! I get your points about the tooltip width and behaviour.

Just one possible quirk I can see – when I click to reveal the tooltip, the "Learn more" link has its accessibility outline.

<img width="529" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/e282f2a2-8a08-4474-91e2-2037b4ca1a0a">

Dan-Q commented:
> Just one possible quirk I can see – when I click to reveal the tooltip, the "Learn more" link has its accessibility outline.

Yeah, I noticed that: I assumed that it was a deliberate (a11y?) choice for this component to focus the contents of the tooltip when opened. I'll double-check!

andfinally commented:
Hi @Dan-Q, just a nudge about this PR! I'm getting a webpack build error locally – I tried to rebase on trunk, but ran into several conflicts.

Dan-Q committed:
Merge branch 'trunk' into add/wccom19063-in-app-shared-by-badge

Dan-Q commented:
> Hi @Dan-Q, just a nudge about this PR! I'm getting a webpack build error locally – I tried to rebase on trunk, but ran into several conflicts.

I just merged changes from trunk without any conflicts, and I'm not getting build errors. What errors are you getting @andfinally?

Dan-Q commented:
Ah! Found the build error. Looks like the underlying TypeScript type doesn't anticipate an actually-valid CSS declaration. Lemme fix that!

Dan-Q committed:
Switch wordWrap for overflowWrap

Modern browesers interpret the former as an alias of the latter, including allowing the use of 'anywhere'. But the current version of https://www.npmjs.com/package/csstype only recognises 'anywhere' as valid for overflowWrap. Switching it prevents compilation errors, still exhibits the correct behavior.

Dan-Q committed:
Merge branch 'trunk' into add/wccom19063-in-app-shared-by-badge

Dan-Q committed:
Prettier linting

Dan-Q commented:
Okay @andfinally! Try that now!

andfinally commented on plugins/woocommerce/changelog/46229-add-wccom19063-in-app-shared-by-badge:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: add
+
+For shared subscriptions, My Subscriptions now shows "Shared with you" and the email address of the person who shared it with you.
```
```suggestion
For shared subscriptions, My Subscriptions now has a "Shared with you" label, with a tooltip showing the email address of the person who shared it with you.
```

andfinally reviewed (approved):
Thanks, works as expected!

I'm still getting the horrible outline I mentioned in https://github.com/woocommerce/woocommerce/pull/46229#pullrequestreview-2006222682. It's coming from a WordPress rule on all links in `common.css`.

I'd suggest we override it.

```diff
Index: plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss b/plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss
--- a/plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss	(revision d627bcdcd532e1f0a34469b9f03f5a6411832c6c)
+++ b/plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss	(date 1715767787120)
@@ -198,6 +198,10 @@
 		color: $gray-900;
 		a {
 			text-decoration: none;
+
+			&:focus {
+				box-shadow: none;
+			}
 		}
 	}
 }
```

I'd even argue `StatusPopover` needs a pointer cursor, as it's interactive.

Dan-Q committed:
Update plugins/woocommerce/changelog/46229-add-wccom19063-in-app-shared-by-badge

Co-authored-by: And Finally <andfinally@users.noreply.github.com>

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Dan-Q committed:
Use pointer as cursor for status popovers

Dan-Q committed:
Merge branch 'add/wccom19063-in-app-shared-by-badge' of https://github.com/woocommerce/woocommerce into add/wccom19063-in-app-shared-by-badge

Dan-Q committed:
Don't focus-within the status badge popover on load (prevents ugly link focus issue)

Dan-Q commented:
> I'm still getting the horrible outline I mentioned in [#46229 (review)](https://github.com/woocommerce/woocommerce/pull/46229#pullrequestreview-2006222682). It's coming from a WordPress rule on all links in `common.css`.
> 
> I'd suggest we override it.

If we override it in the way you suggest, it's bad for a11y. Instead, how about we just don't focus-within the popover when it appears? [2a36975732c92d58f83e908628ebf7303d903345]

https://github.com/woocommerce/woocommerce/assets/53293/abf21b9b-87f1-47c9-aaa5-c7358372a567

andfinally commented:
I'm confused why we're highlighting the link if the focus is on its parent though! And half of me thinks we focus on the tooltip by deafult for accessibility reasons – for example to make it easier to tab through to any link inside it. But this looks like an easy change that resolves the focus issue, thanks!

Dan-Q commented:
> I'm confused why we're highlighting the link if the focus is on its parent.

It looks like the default behaviour for a `<Popover>` is to focus on the first focussable target _within_ it, to make keyboard navigation easier if e.g. you're popping up a toolbar.
